Library/Sender.ts

```diff
@@ -150,38 +156,112 @@ class Sender {
     }
 
     private _confirmDirExists(directory: string, callback: (err: NodeJS.ErrnoException) => void): void {
-        fs.exists(directory, (exists) => {
-            if (!exists) {
+        fs.lstat(directory, (err, stats) => {
+            if (err && err.code === 'ENOENT') {
                 fs.mkdir(directory, (err) => {
                     callback(err);
                 });
-            } else {
+            } else if (!err && stats.isDirectory()){
                 callback(null);
+            } else {
+                callback(err || new Error("Path existed but was not a directory"));
+            }
+        });
+    }
+
+    /**
+     * Computes the size (in bytes) of all files in a directory at the root level. Asynchronously.
+     */
+    private _getShallowDirectorySize(directory: string, callback: (err: NodeJS.ErrnoException, size: number) => void) {
+        // Get the directory listing
+        fs.readdir(directory, (err, files) => {
+            if (err) {
+                return callback(err, -1);
+            }
+
+            let error: NodeJS.ErrnoException = null;
+            let totalSize = 0;
+            let count = 0;
+
+            if (files.length === 0) {
+                callback(null, 0);
+                return;
+            }
+
+            // Query all file sizes
+            for (let i = 0; i < files.length; i++) {
+                fs.stat(path.join(directory, files[i]), (err, fileStats) => {
+                    count++;
+
+                    if (err) {
+                        error = err;
+                    } else {
+                        if (fileStats.isFile()) {
+                            totalSize += fileStats.size;
+                        }
+                    }
+
+                    if (count === files.length) {
+                        // Did we get an error?
+                        if (error) {
+                            callback(error, -1);
+                        } else {
+                            callback(error, totalSize);
+                        }
+                    }
+                });
             }
         });
     }
 
+    /**
+     * Computes the size (in bytes) of all files in a directory at the root level. Synchronously.
+     */
+    private _getShallowDirectorySizeSync(directory: string): number {
+        let files = fs.readdirSync(directory);
+        let totalSize = 0;
+        for (let i = 0; i < files.length; i++) {
+            totalSize += fs.statSync(path.join(directory, files[i])).size;
+        }
+        return totalSize;
+    }
+
     /**
      * Stores the payload as a json file on disk in the temp directory
      */
     private _storeToDisk(payload: any) {
-
-        //ensure directory is created
+        // tmpdir is /tmp for *nix and USERDIR/AppData/Local/Temp for Windows
```

Comment:
Are we sure it's always USERDIR on Windows? I remember there were cases when %temp% defaults to "\Windows\Temp" where many users may have access. If it can end up in Windows\Temp - is there a way in node to give Windows permissions on directory such as ACL to protect it? If it can't - please disregard the comment.

Reply:
It's not guaranteed to always be USERDIR, the system configuration is used to get the temp directory. Here's the logic: https://github.com/nodejs/node/blob/master/lib/os.js#L121
```
path = process.env.TEMP ||
       process.env.TMP ||
       (process.env.SystemRoot || process.env.windir) + '\\temp';
if (path.length > 1 && path.endsWith('\\') && !path.endsWith(':\\'))
  path = path.slice(0, -1);
```

So if the `TEMP` or `TMP` environment variable is changed, that new temp will be respected instead.

As for ACL: In node itself the chmod arguments we're already using is the API presented for changing permissions. Unfortunately the underlying libuv library that does this on node's behalf doesn't map chmod to ACL in a way that will set anything other than the readonly flag.

If we're interested in doing this beyond what node itself supports, we'll need native code(or a library that itself has native code). I found one library that does this, but it doesn't have the same amount of node version support that we have (and I'm not yet sure of the implications of taking a dependency on windows-only native code)